README.md
µTube
=====

### What is µTube?
µTube is an alternative subscription manager for YouTube.

### Getting started
[Click here](http://polyfloyd.github.io/uTube/) for the latest stable live
version. You can also run µTube yourself by cloning it into the public
directory of a webserver.

To add a channel, click in the list-icon thinghy in the top-right (or left
depending on the theme) of the screen and gently input the channel.
Accepted inputs are:
 *  A channelname
 *  A channelID
 *  The URL of the channel

### YouTube API V3
The new V3 API requires all requests to be authorized using a key. Since no such
key is included by default, you will have to create one yourself from the
[Google Developers Console](https://console.developers.google.com/project).
You can change the key in the configuration menu. There is also an option to
change the root URL of the YouTube API should you desire so.
__Finding a creative way to circumvent this limitation is against YouTube's TOS
and should not be attempted.__

### Hotkeys
 *  R = Reload all channels
 *  C = Open the µTube configuration dialog
 *  I = Open the channel menu
 *  Q, Escape = Close a video or dialog
 *  O, Return = Open selected video
 *  H, Left = Move selector left
 *  J, Down = Move selector down
 *  K, Up = Move selector up
 *  L, Right = Move selector right

### Request Policy
µTube needs to get information from the following domains:
 *  ggpht.com (Channel images)
 *  googleapis.com (The YouTube API)
 *  googleusercontent.com (More channel images)
 *  googlevideo.com (Native videos)
 *  youtube-nocookie.com (Embedded videos)
 *  youtube.com (The old YouTube API)
 *  ytimg.com (Video thumbnails images)

### FAQ
 *  What about playlists?

    I have never used YouTube playlists myself, so I have no idea how to
    implement them in a sensible way. Contact me if you want to be the ideas
    guy.

 *  How do I pronounce the name?

    "mewtube"

 *  Will you be adding more stuff in the future?

    Probably, either if I or other people want something.

### Troubleshooting
If something doesn't work, try some of the following:
 *  Use another browser. I only tested this in Firefox and Chromium.
 *  Disable NoScript, Request Policy and whatnot.
 *  Hit the reset button in the config.
 *  Panic.

### License
µTube is available under the MIT license.


CHANGELOG
v 2.2.0
  - Added Piwik analytics.
  - Reorganised items in the configuration menu

v 2.1.0
  - Added the Material theme (SirQuackTheDuck)

v 2.0.2
  - Fixed watched videos not being marked on reload.

v 2.0.1
  - Fixed videos not showing up when using YTAPI2 (off by one)
  - Fixed configuration export

v 2.0.0
  - Use asynchronous XMLHttpRequests, improving loading speed
  - Added a part of the video description to video tooltips
  - Added a config option to control the amount of videos loaded
  - Channels can now be added by name, ID and URL
  - Fixed a few graphical quirks
  - Added support for YouTube Native playback
  - Added support for YouTube API V3
  - Drop native playback using ViewTube. Use YouTube HTML5 instead.
  - Added an option to disable the alternative horizontal scrolling

v 1.0.4
  - Show hours and minutes on recent videos

v 1.0.3
  - Remove theme comments
  - The browser is now in charge of selecting the proper video format

v 1.0.2
  - Dropped the Mickeysoft theme
  - Removed the text-shadow from the Dusk theme
  - Fixed some crappy CSS
  - Update ViewTube module

v 1.0.1
  - Fixed channel titles being displayed incorrectly in Chromium

v 1.0
  - Inital release
  - Subscribe to YouTube channels without the need for a Google+ account


js/assets.js
var FOOTER_MAX = 60;
var FOOTER_MIN = 8;

function channelbox() {
	return document.querySelector("div.ut_channelbox");
}

function footer() {
	return document.querySelector("div.ut_footer");
}

function showFooter() {
	channelbox().style.height = "calc(100% - " + (122 + FOOTER_MAX) + "px)";
	footer().style.height = FOOTER_MAX + "px";
}

function hideFooter() {
	channelbox().style.height = "calc(100% - " + (122 + FOOTER_MIN) + "px)";
	footer().style.height = FOOTER_MIN + "px";
}

function updatetransitionConf(enable) {
	var style = document.getElementById("disabletransitions");
	if (!enable && !style) {
		style = document.createElement("style");
		style.id = "disabletransitions";
		style.setAttribute("type", "text/css");
		style.innerHTML = "*{transition:none !important;}";
		document.getElementsByTagName("head")[0].appendChild(style);
	} else if (style) {
		style.remove();
	}
}

function addChannelByForm() {
	var inputElem = document.getElementsByClassName("ut_addchannel_txt")[0];
	var oldErrs = document.getElementsByClassName("ut_chanconf_err");
	for (var i = 0; i < oldErrs.length; i++) {
		oldErrs[i].remove();
	}
	var err = utube.chan.add(inputElem.value);
	if (err) {
		var errElem = document.createElement("h6");
		errElem.classList.add("ut_chanconf_err");
		errElem.innerHTML = err;
		inputElem.parentNode.appendChild(errElem);
	} else {
		utube.updateChannelMenu();
		utube.updateChannels();
	}
}

function removeChannelByForm(name) {
	utube.chan.remove(name);
	utube.updateChannels();
}


js/youtube.js
/**
 *  uTube, Tube without the Google
 *  Copyright (C) 2013-2015  The uTube authors
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

'use strict';

function getJSON(url, cb) {
  var req = new XMLHttpRequest();
  req.onreadystatechange = function() {
    if (req.readyState !== XMLHttpRequest.DONE) return;
    var data = null;
    try {
      data = JSON.parse(req.responseText);
    } catch(err) {
      cb(err, null);
      return;
    }
    cb(null, data);
  };
  req.open('GET', url, true);
  req.send();
};

function inherit(sup, sub) {
  for (var k in sup.prototype) sub.prototype[k] = sup.prototype[k];
  return sub;
}

var GenericAPI = {
  Channel: function() {},
  Video:   function() {},
};

GenericAPI.Channel.prototype.idIsID = function() {
  return !!this.id.match(/^[a-zA-Z0-9_-]{24}$/);
};


GenericAPI.Channel.prototype.getLink = function() {
  var url = 'https://www.youtube.com/{{ selector }}/{{ id }}/featured';
  return url.template({
    id:       this.id,
    selector: this.idIsID() ? 'channel' : 'user',
  });
};

GenericAPI.Video.prototype.getThumb = function(type) {
  type = type || 'def';
  return 'https://i1.ytimg.com/vi/{{ id }}/{{ type }}.jpg'.template({
    id:   this.id,
    type: {
      def: 'mqdefault',
      max: 'maxresdefault',
    }[type],
  });
};

GenericAPI.Video.prototype.getLink = function(options) {
  return 'http://www.youtube.com/watch?v={{ id }}&html5={{ html5 }}'.template({
    html5: options.html5 ? '1' : '0',
    id:    this.id,
  });
};

GenericAPI.Video.prototype.getEmbedLink = function(options) {
  return 'https://www.youtube-nocookie.com/embed/{{ id }}?autoplay={{ autoplay }}&html5={{ html5 }}'.template({
    autoplay: options.autoplay ? '1' : '0',
    html5:    options.html5    ? '1' : '0',
    id:       this.id,
  });
};

var YouTubeAPI2 = {};

YouTubeAPI2.Channel = inherit(GenericAPI.Channel, function(options) {
  this.icon  = options.icon;
  this.id    = options.id;
  this.title = options.title;
});

YouTubeAPI2.Video = inherit(GenericAPI.Video, function(options) {
  this.description = options.description;
  this.duration    = options.duration;
  this.id          = options.id;
  this.published   = options.published;
  this.title       = options.title;
});

YouTubeAPI2.Channel.prototype.getVideos = function(next, limit, cb) {
  var url = 'https://gdata.youtube.com/feeds/api/users/{{ id }}/uploads?alt=json&orderby=published&start-index={{ offset }}&max-results={{ limit }}';
  var offset = next ? parseInt(next, 10) + 1 : 1;
  getJSON(url.template({
    id:     this.id,
    limit:  limit,
    offset: offset,
  }), function(err, data) {
    if (err) {
      cb(err, null);
      return;
    }
    data.feed.entry = data.feed.entry || [];
    var videos = data.feed.entry.map(function(video) {
      return new YouTubeAPI2.Video({
        description: video.media$group.media$description.$t,
        duration:    parseInt(video.media$group.yt$duration.seconds, 10),
        id:          function(id) {
          return id.substring(id.lastIndexOf('/') + 1, id.length);
        }(video.id.$t),
        published:   new Date(video.published.$t),
        title:       video.title.$t,
      });
    });
    cb(err, err ? null : {
      next:   offset + limit - 1,
      videos: videos,
    });
  });
};

YouTubeAPI2.getChannelByID = function(id, cb) {
  var url = 'https://gdata.youtube.com/feeds/api/users/{{ id }}?alt=json';
  getJSON(url.template({ id: id }), function(err, data) {
    cb(err, err ? null : new YouTubeAPI2.Channel({
      icon:  data.entry.media$thumbnail.url,
      id:    id,
      title: data.entry.title.$t,
    }));
  });
};

var YouTubeAPI3 = {};

YouTubeAPI3.key = '';
YouTubeAPI3.url = '';

YouTubeAPI3.Channel = inherit(GenericAPI.Channel, function(options) {
  this.icon    = options.icon;
  this.id      = options.id;
  this.title   = options.title;
  this.uploads = options.uploads;
});

YouTubeAPI3.Video = inherit(GenericAPI.Video, function(options) {
  this.description = options.description;
  this.duration    = options.duration;
  this.id          = options.id;
  this.published   = options.published;
  this.title       = options.title;
});

YouTubeAPI3.Channel.prototype.getVideos = function(next, limit, cb) {
  var url = '{{ root }}playlistItems?part=id%2Csnippet&maxResults={{ limit }}&pageToken={{ next }}&playlistId={{ id }}&key={{ key }}';
  getJSON(url.template({
    id:    this.uploads,
    key:   YouTubeAPI3.key,
    limit: limit,
    next:  next || '',
    root:  YouTubeAPI3.url,
  }), function(err, data) {
    if (err) {
      cb(err, null);
      return;
    }
    var videos = data.items.map(function(video) {
      return new YouTubeAPI3.Video({
        description: video.snippet.description,
        duration:    0, // TODO
        id:          video.snippet.resourceId.videoId,
        published:   new Date(video.snippet.publishedAt),
        title:       video.snippet.title,
      });
    });
    cb(err, err ? null : {
      next:   data.nextPageToken,
      videos: videos,
    });
  });
};

YouTubeAPI3.getChannelByID = function(id, cb) {
  var url = '{{ root }}channels?part=id%2Csnippet%2CcontentDetails&{{ selector }}={{ id }}&key={{ key }}';
  getJSON(url.template({
    id:       id,
    key:      YouTubeAPI3.key,
    root:     YouTubeAPI3.url,
    selector: GenericAPI.Channel.prototype.idIsID.call({ id: id }, {}) ? 'id' : 'forUsername',
  }), function(err, data) {
    if (!data.items[0]) {
      cb(null, null)
      return;
    }
    cb(err, err ? null : new YouTubeAPI3.Channel({
      icon:    data.items[0].snippet.thumbnails.default.url,
      id:      data.items[0].id,
      title:   data.items[0].snippet.title,
      uploads: data.items[0].contentDetails.relatedPlaylists.uploads,
    }));
  });
};


js/main.js
/**
 *  uTube, Tube without the Google
 *  Copyright (C) 2013-2015  The uTube authors
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

'use strict';

Date.prototype.format = function(fmt) {
  function pad(n, minLength) {
    var str = ''+n;
    for (; str.length < minLength; str = '0'+str);
    return str;
  }
  fmt = fmt.replace('%ms', this.getMilliseconds());
  fmt = fmt.replace('%s', this.getSeconds());
  fmt = fmt.replace('%m', pad(this.getMinutes(), 2));
  fmt = fmt.replace('%h', pad(this.getHours(), 2));
  fmt = fmt.replace('%d', this.getDate());
  fmt = fmt.replace('%M', this.getMonth() + 1);
  fmt = fmt.replace('%Y', this.getFullYear());
  return fmt;
};

String.prototype.template = function(args) {
  var str = this;
  for (var k in args) {
    str = str.replace(new RegExp('\\{\\{ '+k+' \\}\\}', 'g'), args[k]);
  }
  return str;
};

Array.prototype.remove = function(what) {
  var i;
  while ((i = this.indexOf(what)) !== -1) {
    this.splice(i, 1);
  }
  return this;
};

Node.prototype.remove = Node.prototype.remove || function() {
  this.parentNode.removeChild(this);
};

Node.prototype.removeAll = function() {
  while (this.hasChildNodes()) {
    this.removeChild(this.lastChild);
  }
};

HTMLCollection.prototype.toArray = function() {
  return Array.prototype.slice.call(this);
};

EventTarget.prototype.addMousewheel = function(listener) {
  this.addEventListener('mousewheel',     listener, false);
  this.addEventListener('DOMMouseScroll', listener, false);
};

EventTarget.prototype.removeMousewheel = function(listener) {
  this.removeEventListener('mousewheel',     listener, false);
  this.removeEventListener('DOMMouseScroll', listener, false);
};

// Meh, who needs jQuery anyway?
window.$ = function() {
  return document.querySelector.apply(document, arguments);
};

var VERSION = '2.2.0';

var utube = {

  conf: {

    onchange: {},

    reset: function(source) {
      source = source || utube.conf.standard;
      localStorage.clear();
      for (var key in source) {
        utube.conf.set(key, source[key], true);
      }
      window.location.reload(false);
    },

    get: function(key) {
      try {
        return JSON.parse(localStorage.getItem(key));
      } catch (err) {
        return null;
      }
    },

    set: function(key, value, noapply) {
      localStorage.setItem(key, JSON.stringify(value));
      if (!noapply && utube.conf.onchange[key]) utube.conf.fireChanged(key, value);
    },

    fireChanged: function(key, value) {
      if (typeof value === 'undefined') value = utube.conf.get(key);
      utube.conf.onchange[key](value);
    },

    exportAll: function() {
      var temp = {};
      for (var k in utube.conf.standard) {
        temp[k] = utube.conf.get(k);
      }
      return JSON.stringify(temp);
    },

    importAll: function(data) {
      if (!data) return true;
      try {
        utube.conf.reset(JSON.parse(data));
        return true;
      } catch (err) {
        return false;
      }
    },

    standard: {
      althscroll:     true,
      analytics:      true,
      api:            'YouTubeAPI2',
      apiyt3key:      '',
      apiyt3url:      'https://www.googleapis.com/youtube/v3/',
      autoplay:       true,
      channels:       [],
      chanorder:      'VIDDATE',
      loadincrement:  6,
      markwatched:    true,
      onvidclick:     'EMBED',
      playback:       'HTML5',
      theme:          'dusk.css',
      transitions:    true,
      version:        VERSION,
    },

    themes: [
      {
        name:   'Dusk',
        source: 'dusk.css',
      },
      {
        name:   'New Horizons',
        source: 'newhorizons.css',
      },
      {
        name:   'Oh Es Icks',
        source: 'ohesicks.css',
      },
      {
        name:   'Material',
        source: 'material/theme.css',
      }
    ],

  },

  chan: {

    getAll: function() {
      return utube.conf.get('channels').map(function(chan) {
          return new utube.yt.Channel(chan);
      });
    },

    exists: function(channel) {
      return utube.chan.getAll().some(function(chan) {
        return chan.id === channel.id;
      });
    },

    add: function(channel, cb) {
      if (utube.chan.exists(channel)) {
        cb(new Error(channel.title+' is already added'));
        return;
      }
      var channels = utube.chan.getAll();
      channels.push(channel);
      channels.sort(function(a, b) {
        return a.title > b.title;
      });
      utube.conf.set('channels', channels);
      cb(null);
    },

    remove: function(channel) {
      utube.conf.set('channels', utube.chan.getAll().filter(function(chan) {
        return chan.id !== channel.id;
      }));
    },

  },

  showConfigMenu: function() {
    var menu = document.createElement('div');
    menu.classList.add('ut_configmenu');
    menu.style.width = '500px';
    menu.innerHTML = $('#ut_configmenu_content').innerHTML.template({
      version: VERSION,
    });
    utube.showOverlay(menu);
    var themeSelect = menu.querySelector('select');
    utube.conf.themes.forEach(function(theme) {
      var op = document.createElement('option');
      op.value     = theme.source;
      op.innerHTML = theme.name;
      themeSelect.appendChild(op);
    });
    menu.getElementsByTagName('input').toArray().forEach(function(input) {
      switch (input.type) {
        case 'text':
          input.value = utube.conf.get(input.name);
          input.setAttribute('onchange', 'utube.conf.set(this.name, this.value)');
          break;
        case 'number':
          input.value = utube.conf.get(input.name);
          input.setAttribute('onchange', 'utube.conf.set(this.name, parseInt(this.value))');
          break;
        case 'checkbox':
          if (utube.conf.get(input.name)) {
            input.checked = 'checked';
          }
          input.setAttribute('onclick', 'utube.conf.set(this.name, !!this.checked)');
          break;
        case 'radio':
          if (utube.conf.get(input.name) == input.value) {
            input.checked = 'checked';
          }
          input.setAttribute('onchange', 'utube.conf.set(this.name, this.value)');
          break;
      }
    });
    menu.getElementsByTagName('select').toArray().forEach(function(select) {
      select.setAttribute('onchange', 'utube.conf.set(this.name, this.value)');
      var value =  utube.conf.get(select.name);
      for (var j = select.childNodes.length - 1; j >= 0; j--) {
        var op = select.childNodes[j];
        if (op.value == value) {
          op.setAttribute('selected', 'selected');
          break;
        }
      }
    });
  },

  showChannelMenu: function() {
    var menu = document.createElement('div');
    menu.style.width = '500px';
    menu.classList.add('ut_channelmenu');
    menu.innerHTML = $('#ut_channelmenu_content').innerHTML;
    var list = menu.querySelector('.ut_channelmenu_list');
    list.style.height = (document.documentElement.clientHeight - 200)+'px';
    utube.showOverlay(menu);
    utube.updateChannelMenu();
  },

  updateChannelMenu: function() {
    var itemTemplate = $('#ut_channelmenu_item').innerHTML;
    $('.ut_channelmenu_list').innerHTML = utube.chan.getAll().reduce(function(prev, channel) {
      return prev+itemTemplate.template(channel);
    }, '');
  },

  addChannelByForm: function() {
    var inputElem = $('.ut_addchannel_txt');
    var oldErr    = $('.ut_chanconf_err');
    if (oldErr) oldErr.remove();

    function error(err) {
      var errElem = document.createElement('h6');
      errElem.classList.add('ut_chanconf_err');
      errElem.innerHTML = err.message;
      inputElem.parentNode.appendChild(errElem);
    }

    var string = [
      /^([a-zA-Z0-9]{1,20})$/,
      /^([a-zA-Z0-9_-]{24})$/,
      /^https?:\/\/www\.youtube\.com\/user\/([a-zA-Z0-9]{1,20}).*/,
      /^https?:\/\/www\.youtube\.com\/channel\/([a-zA-Z0-9_-]{24}).*/,
    ].reduce(function(prev, regex) {
      var m = inputElem.value.match(regex);
      return prev || (m ? m[1] : null);
    }, null);
    if (!string) {
      error(new Error('No valid channel identifier not found in input'));
      return;
    }

    utube.yt.getChannelByID(string, function(err, channel) {
      if (err) {
        error(err);
        return;
      }
      if (!channel) {
        error(new Error('Channel not found'));
        return;
      }
      utube.chan.add(channel, function(err) {
        if (err) {
          error(err);
          return;
        }
        utube.updateChannelMenu();
        utube.updateChannels();
      });
    });
  },

  removeChannelByForm: function(id) {
    utube.chan.remove(new utube.yt.Channel({ id: id }));
    utube.updateChannels();
  },

  insertVideos: function(channelID, next, limit, vidListElem, cb) {
    utube.chan.getAll().filter(function(chan) {
      return chan.id === channelID;
    })[0].getVideos(next, limit, function(err, result) {
      if (err) {
        var errElem = document.createElement('p');
        errElem.innerHTML = err.message;
        vidListElem.appendChild(errElem);
        console.error(err);
        cb()
        return;
      }
      result.videos.forEach(function(video) {
        var vidElem = document.createElement('div');
        vidElem.id = 'vid_'+video.id;
        vidElem.classList.add('ut_list_video');
        if (utube.hasWatched(video.id)) {
          vidElem.classList.add('ut_video_watched');
        }
        vidElem.setAttribute('onclick', 'utube.playVideo(\''+video.id+'\')');
        vidElem.title = video.title+'\n'+video.description;
        var timeFormat = video.published.getTime() > (new Date().getTime() - 1000 * 60 * 60 * 24)
          ? '%h:%m' : '%d-%M-%Y';
        vidElem.innerHTML = (
          '<p class="ut_video_title">{{ title }}</p>'+
          '<p class="ut_video_time">{{ time }}</p>'+
          '<p class="ut_video_duration">{{ duration }}</p>'+
          '<img src="{{ thumb }}" />'
        ).template({
          duration: utube.timeString(video.duration),
          thumb:    video.getThumb(),
          time:     video.published.format(timeFormat),
          title:    video.title,
        });
        vidListElem.appendChild(vidElem);
      });
      if (!next) {
        var mostRecent = result.videos[0] || { published: new Date(0) };
        vidListElem.parentNode.setAttribute('data-mostrecent', mostRecent.published.getTime());
      }
      vidListElem.setAttribute('data-nextpage', result.next);
      cb();
    });
  },

  updateChannels: function() {
    var channels = utube.chan.getAll();
    utube.selectorX = utube.selectorY = 0;
    utube.selectorYSave = [];
    utube.chan.count = channels.length;
    var chanBox = $('.ut_channelbox');
    chanBox.removeAll();
    channels.forEach(function(channel) {
      var chanElem = document.createElement('div');
      var vidListElem = document.createElement('div');
      vidListElem.classList.add('ut_channel_videos');
      vidListElem.setAttribute('data-channelid', channel.id);
      vidListElem.addMousewheel(utube.scrollVideos);
      chanElem.classList.add('ut_channel');
      chanElem.innerHTML = (
        '<a href="{{ url }}" target="_blank" title="{{ title }}">'+
          '<div class="ut_channel_head">'+
            '<img src="{{ icon }}" />'+
            '<h3>{{ title }}</h3>'+
          '</div>'+
        '</a>'
      ).template({
        icon:  channel.icon,
        title: channel.title,
        url:   channel.getLink(),
      });
      chanElem.appendChild(vidListElem);
      chanBox.appendChild(chanElem);
      var increment = utube.conf.get('loadincrement');
      utube.insertVideos(channel.id, null, increment, vidListElem, function() {
        if (utube.conf.get('chanorder') === 'VIDDATE') {
          document.getElementsByClassName('ut_channel').toArray().sort(function(a, b) {
            var ta = parseInt(a.getAttribute('data-mostrecent'), 10);
            var tb = parseInt(b.getAttribute('data-mostrecent'), 10);
            return ta < tb;
          }).forEach(function(channel) {
            channel.parentNode.appendChild(channel);
          });
        }
      });
    });
  },

  playVideo: function(id) {
    var video = new utube.yt.Video({ id: id });
    function getEmbeddedVideo() {
      return video.getEmbedLink({
        autoplay: utube.conf.get('autoplay'),
        html5:    utube.conf.get('playback') === 'HTML5',
      });
    }
    switch (utube.conf.get('onvidclick')) {
      case 'EMBED':
        var embedElem = document.createElement('iframe');
        embedElem.classList.add('ut_embedvideo');
        embedElem.src = getEmbeddedVideo();
        embedElem.setAttribute('allowfullscreen', 'allowfullscreen');
        var height = window.innerHeight - 100;
        var width  = window.innerWidth  - 100;
        var w = height * (16 / 9);
        var h = height;
        if (w > width) {
          w = width;
          h = width / (16 / 9);
        }
        embedElem.width = w;
        embedElem.height = h;
        utube.showOverlay(embedElem);
        break;
      case 'EMBEDINTAB':
        window.open(getEmbeddedVideo());
        break;
      case 'OPENYT':
        window.open(video.getLink({
          html5: utube.conf.get('playback') === 'HTML5',
        }));
        break;
    }
    if (utube.conf.get('markwatched')) {
      utube.markAsWatched(video.id);
    }
  },

  markAsWatched: function(id) {
    utube.conf.set('watched_'+id, true);
    $('#vid_'+id).classList.add('ut_video_watched');
  },

  hasWatched: function(id) {
    return utube.conf.get('watched_'+id);
  },

  unwatchAll: function() {
    for (var k in localStorage) {
      if (k.indexOf('watched_') === 0) {
        localStorage.removeItem(k);
      }
    }
    document.getElementsByClassName('ut_video_watched').toArray().forEach(function(v) {
      v.classList.remove('ut_video_watched');
    });
  },

  isOverlayOpen: function() {
    return !!$('.ut_overlay');
  },

  showOverlay: function(contentElem) {
    if (utube.isOverlayOpen()) return false;
    var ov = document.createElement('table');
    var tr = document.createElement('tr');
    var td = document.createElement('td');
    var wr = document.createElement('div');
    wr.classList.add('ut_overlay_wrapper');
    contentElem.classList.add('ut_overlay_content');
    contentElem.onclick = function(e) { e.stopPropagation(); };
    wr.appendChild(contentElem);
    td.appendChild(wr);
    tr.appendChild(td);
    ov.appendChild(tr);
    ov.classList.add('ut_overlay');
    ov.style.opacity = '0';
    td.onclick = utube.removeOverlay;
    contentElem.getElementsByTagName('input').toArray().forEach(function(input) {
      if (input.type === 'text') {
        input.setAttribute('onfocus', 'utube.keysEnabled=false');
        input.setAttribute('onblur', 'utube.keysEnabled=true');
      }
    });
    setTimeout(function() {
      ov.style.opacity = '1';
    }, utube.conf.get('transitions') ? 20 : 0);
    var body = $('body');
    body.insertBefore(ov, body.childNodes[1]);
    wr.style.width = contentElem.clientWidth+'px';
  },

  removeOverlay: function() {
    var ov = $('.ut_overlay');
    if (ov) {
      ov.style.opacity = '0';
      setTimeout(function() {
        ov.remove();
      }, utube.conf.get('transitions') ? 300 : 0);
    }
  },

  loadVideos: function(vidList) {
    if (!vidList.classList.contains('ut_loading')) {
      vidList.classList.add('ut_loading');
      var increment = utube.conf.get('loadincrement')
      var next      = vidList.getAttribute('data-nextpage');
      utube.insertVideos(vidList.getAttribute('data-channelid'), next, increment, vidList, function() {
        vidList.classList.remove('ut_loading');
      });
    }
  },

  scrollVideos: function(e) {
    var n = e.target;
    for (; n && !n.classList.contains('ut_channel_videos'); n = n.parentNode);
    if (n) {
      n.scrollTop -= e.wheelDelta / 2 || -e.detail * 20;
      if (n.scrollTop == n.scrollHeight - n.clientHeight) {
        utube.loadVideos(n);
      }
    }
    e.stopPropagation();
  },

  timeString: function(seconds) {
    seconds = parseInt(seconds);
    var s = '';
    var hasHours = seconds > 3600;
    if (hasHours) {
      s += Math.round(seconds / 3600)+':';
      seconds %= 3600;
    }
    var min = Math.round(seconds / 60 - 0.5);
    if (min < 10 && hasHours) {
      s += '0';
    }
    s += min+':';
    var sec = seconds % 60;
    if (sec < 10) {
      s += '0';
    }
    return s + sec;
  },

  updateSelector: function() {
    var old = $('.ut_list_video.ut_selected');
    if (old) {
      old.classList.remove('ut_selected');
    }
    var vid = utube.getSelectedVideo();
    vid.classList.add('ut_selected');
    vid.scrollIntoView(false);
    if (vid == old) {
      utube.loadVideos(vid.parentNode);
    }
  },

  playSelectedVideo: function() {
    if (!utube.isOverlayOpen()) {
      utube.getSelectedVideo().onclick();
    }
  },

  getSelectedChannel: function() {
    var cbox = $('.ut_channelbox');
    return cbox.childNodes[utube.selectorX];
  },

  getSelectedVideo: function() {
    var chan = utube.getSelectedChannel();
    if (chan) {
      var list = chan.querySelector('.ut_channel_videos');
      return list.childNodes[utube.selectorY];
    }
    return null;
  },

  selectorMoveUp: function() {
    if (utube.selectorY > 0 && !utube.isOverlayOpen()) {
      utube.selectorY--;
      utube.updateSelector();
    }
  },

  selectorMoveDown: function() {
    if (!utube.isOverlayOpen()) {
      var vidcount = utube.getSelectedChannel().getElementsByClassName('ut_list_video').length;
      if (utube.selectorY < vidcount - 1) {
        utube.selectorY++;
      }
      utube.updateSelector();
    }
  },

  selectorMoveLeft: function() {
    if (utube.selectorX > 0 && !utube.isOverlayOpen()) {
      utube.selectorYSave[utube.selectorX] = utube.selectorY;
      utube.selectorX--;
      utube.selectorY = utube.selectorYSave[utube.selectorX] || 0;
      utube.updateSelector();
    }
  },

  selectorMoveRight: function() {
    if (utube.selectorX < utube.chan.count - 1 && !utube.isOverlayOpen()) {
      utube.selectorYSave[utube.selectorX] = utube.selectorY;
      utube.selectorX++;
      utube.selectorY = utube.selectorYSave[utube.selectorX] || 0;
      utube.updateSelector();
    }
  },

  addKey: function(key, callback) {
    utube.keybindings[key] = callback;
  },

  alternativeHorizontalScroll: function(e) {
    $('.ut_channelbox').scrollLeft -= e.wheelDelta / 2 || -e.detail * 20;
  },

  init: function() {
    if (!utube.conf.get('version')) {
      utube.conf.reset();
      return;
    }
    if (utube.conf.get('version') <= '2.1.0') {
      utube.conf.set('analytics', true);
    }
    document.addEventListener('keydown', function(e) {
      var func = utube.keybindings[e.keyCode];
      if (func && utube.keysEnabled) func(e);
    }, false);
    utube.keybindings = {};
    utube.keysEnabled = true;
    utube.addKey(82, utube.updateChannels);    // R
    utube.addKey(67, utube.showConfigMenu);    // C
    utube.addKey(73, utube.showChannelMenu);   // I
    utube.addKey(81, utube.removeOverlay);     // Q
    utube.addKey(79, utube.playSelectedVideo); // O
    utube.addKey(75, utube.selectorMoveUp);    // K
    utube.addKey(74, utube.selectorMoveDown);  // J
    utube.addKey(72, utube.selectorMoveLeft);  // H
    utube.addKey(76, utube.selectorMoveRight); // L
    utube.addKey(27, utube.removeOverlay);     // Escape
    utube.addKey(38, utube.selectorMoveUp);    // Up
    utube.addKey(40, utube.selectorMoveDown);  // Down
    utube.addKey(37, utube.selectorMoveLeft);  // Left
    utube.addKey(39, utube.selectorMoveRight); // Right
    utube.addKey(13, utube.playSelectedVideo); // Return
    utube.conf.onchange.chanorder   = utube.updateChannels;
    utube.conf.onchange.theme       = function() {
      $('.ut_theme').setAttribute('href', 'css/theme/'+utube.conf.get('theme'));
    };
    utube.conf.onchange.althscroll = function(enable) {
      if (enable) {
        $('.ut_channelbox').addMousewheel(utube.alternativeHorizontalScroll);
        $('.ut_cbar').addMousewheel(utube.alternativeHorizontalScroll);
      } else {
        $('.ut_channelbox').removeMousewheel(utube.alternativeHorizontalScroll);
        $('.ut_cbar').removeMousewheel(utube.alternativeHorizontalScroll);
      }
    };
    utube.conf.onchange.analytics = function(enable) {
      if (enable) {
        window._paq = window._paq || [];
        window._paq.push(['trackPageView']);
        window._paq.push(['enableLinkTracking']);
        (function(_paq) {
          var u="//analytics.polyfloyd.net/";
          _paq.push(['setTrackerUrl', u+'piwik.php']);
          _paq.push(['setSiteId', 3]);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'piwik.js'; s.parentNode.insertBefore(g,s);
        })(_paq);
      }
    };
    utube.conf.onchange.transitions = function(enable) {
      var style = $('#disabletransitions');
      if (!enable && !style) {
        style = document.createElement('style');
        style.id = 'disabletransitions';
        style.setAttribute('type', 'text/css');
        style.innerHTML = '*{transition:none !important;}';
        $('head').appendChild(style);
      } else if (style) {
        style.remove();
      }
    };
    utube.conf.onchange.api = function(api) {
      utube.yt = {
        YouTubeAPI2: YouTubeAPI2,
        YouTubeAPI3: YouTubeAPI3,
      }[api];
      if (api === 'YouTubeAPI3') {
        utube.conf.fireChanged('apiyt3key');
        utube.conf.fireChanged('apiyt3url');
        utube.chan.getAll().filter(function(channel) {
          return !channel.uploads;
        }).forEach(function(channel) {
          utube.yt.getChannelByID(channel.id, function(err, newChannel) {
            if (!err && newChannel) {
              utube.chan.remove(channel);
              utube.chan.add(newChannel, function() {});
            }
          });
        });
      }
    };
    utube.conf.onchange.apiyt3key = function(key) {
      YouTubeAPI3.key = key;
    };
    utube.conf.onchange.apiyt3url = function(url) {
      YouTubeAPI3.url = url;
    };
    [
      'althscroll',
      'analytics',
      'api',
      'theme',
      'transitions',
    ].forEach(function(k) { utube.conf.fireChanged(k); });
    utube.updateChannels();
  },

};
